### PR TITLE
Disable service on release network

### PR DESCRIPTION
**Pull request by abhi**
This PR contains a fix for moby/moby#30321. There was a moby/moby#31142
PR intending to fix the issue by adding a delay between disabling the
service in the cluster and the shutdown of the tasks. However
disabling the service was not deleting the service info in the cluster.
Added a fix to delete service info from cluster and verified using siege
to ensure there is zero downtime on rolling update of a service.In order
to support it and ensure consitency of enabling and disable service knob
from the daemon, we need to ensure we disable service when we release
the network from the container. This helps in making the enable and
disable service less racy. The corresponding part of libnetwork fix is
part of docker/libnetwork#1824

Signed-off-by: abhi <abhi@docker.com>

**- What I did**
In order to ensure enable and disable a service on an sandbox less racy , explictly disable service on the sandbox during a container exit event on the network side. This entry point is the releaseNetwork action which triggers the tearing of the network resource for a container.
**- How I did it**
Added DisableService action on releaseNetwork
**- How to verify it**
This is particularly important during rolling update of a service and when the tasks in a swarm service goes down. Verified using siege and performing zero downtime rolling update.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by abhi** [comment]
@fcrisciani

**Comment by thaJeztah** [comment]
ping @abhi can you rebase this PR? Also looks like you added a file that shouldn't be there (I noticed theres a stack dump in there);

screen shot 2018-01-16 at 11 19 26

**Comment by thaJeztah** [comment]
Looks like you updated the PR; note that GitHub doesn't send out notifications when you update, so its best to comment or "ping" someone after you did so 😄

**Comment by abhi** [comment]
@thaJeztah the guy I need to ping sits next to me :D Will do next time. 
@fcrisciani PTAL

File: daemon/container_operations.go [hunk 1]
```diff
@@ -966,6 +966,9 @@ func (daemon *Daemon) releaseNetwork(container *container.Container) {
 		logrus.Warnf("error locating sandbox id %s: %v", sid, err)
 		return
 	}
+	if err := sb.DisableService(); err != nil {
+		logrus.Errorf("Error disabling service for sandbox id  %s container %s: %v", sid, container.ID, err)
```
Comment (cpuguy83) [inline daemon/container_operations.go]: Nit: `logrus.WithField("container", container.ID).WithError(err).WithField("sandbox", sid).Error("Error removing service from sandbox")`

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: daemon/container_operations.go (hunk 1 above)
Comment (thaJeztah) [reply daemon/container_operations.go]: Or `.WithFields()`;

```go
log.WithFields(log.Fields{"container": container.ID, "sandbox": sid}).WithError(err).Error("Error removing service from sandbox")
```

**Comment by fcrisciani** [comment]
LGTM after the log change 👍

**Comment by abhi** [comment]
@thaJeztah @cpuguy83 done.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by cpuguy83 — APPROVED** [review]
LGTM

I'm not sure I like the ambiguity of the "service" term, but it is what it is.

**Pull request merged**